### PR TITLE
[codex] Split installed-consumer contract helpers

### LINKED ISSUE #312: Partition installed-consumer contract CMake script by responsibility
Parent: #302

Related: #307, #311, #48, #90, #99, #103, #106, #156, #201

## Problem

The supplemental code-quality audit found that the installed-consumer contract script is valuable but overloaded. Install verification, source rejection probes, version probes, launcher logic, and stderr diagnostics share one large CMake script, increasing the chance that package-contract edits disturb unrelated checks.

## Observed evidence

- `tests/check_installed_package_consumer.cmake` combines multiple responsibilities in one script.
- Installed-package behavior is central to the release claim-evidence work in #304 and public-surface change-amplification work in #307.

## Hypothesis

Partitioning helper logic into responsibility-focused included CMake files can improve review locality while preserving the release-contract test surface.

## Validation / acceptance criteria

- First perform an include-file extraction only; do not change the tested contract.
- Preserve test names, pass/fail outcomes, failure diagnostics, and feature-mode coverage.
- Compare CTest output before and after.
- Roll back if the partition makes the CMake flow harder to follow or weakens the installed-consumer contract.

## Scope boundary

Do not remove install-consumer coverage or narrow package checks to reduce file size. This is organization only.

## Risks / trade-offs

CMake include-file sprawl can be worse than one long script. Split only along real responsibilities with stable names.

## Related review outputs

Code-Quality Auditor, Maintainer-Economics Auditor, Interop and Ecosystem Auditor.

### PR DESCRIPTION
**Pull request by jaharris87**
Closes #312.

## Summary

- Keep `tests/check_installed_package_consumer.cmake` as the ordered installed-consumer contract flow.
- Move responsibility-specific helper bodies into `tests/installed-consumer-contract/` includes for setup, producer install, artifact verification, package-version probes, source rejection probes, consumer build, and consumer execution/report diagnostics.
- Preserve the tested contract surface: no install-consumer coverage, version probes, source rejection probes, launcher logic, or diagnostics were removed.

## Validation

- Baseline before edits: `ctest --test-dir build-smoke --output-on-failure -R '^(ftimer_installed_package_consumer|ftimer_absolute_includedir_destdir_install)'` passed 5/5.
- After edits: `ctest --test-dir build-smoke --output-on-failure -R '^(ftimer_installed_package_consumer|ftimer_absolute_includedir_destdir_install)'` passed 5/5.
- `ctest --test-dir build-smoke --output-on-failure -R '^(ftimer_configure_contracts|ftimer_make_wrapper_contracts)$'` passed 2/2.
- `git diff --check` and `git diff --cached --check` passed.
- Mechanical check: concatenating the extracted include bodies reconstructs the original script command text, aside from include-boundary blank separator lines and the new helper-file headers.

### CODE REVIEW

**Comment by jaharris87** [comment]
@codex review as a senior software engineer reviewing this fTimer PR for correctness; inspect changed code, changed tests, and changed user-facing docs, then find behavior regressions, wrapper or OOP parity drift, install or export breakage, unsafe defaults, silent lossy behavior, summary or output regressions, and missing end-to-end validation, not style issues; when timer runtime paths change, explicitly sanity-check start or stop ordering, context attribution, and docs/tests parity; classify findings as bug, test gap, design concern, or nit; begin with "## Software Review".

**Comment by jaharris87** [comment]
@codex review as a skeptical fTimer test reviewer; decide whether the changed tests would fail for broken start or stop ordering, repair timestamp consistency, callback suppression during repair, self-time sibling boundaries, error-contract mistakes, MPI mismatch handling, and worker-thread no-op behavior, and whether mocks or golden tests create false confidence; classify findings as test gap, weak assertion, false confidence, flaky risk, or design concern; begin with "## Test Quality Review".

**Comment by jaharris87** [comment]
Manual fallback review record

Reviewed PR head SHA: `b0c79c6e959a9f1915900cfd7b2215e1e45d78da`
Fallback wave: `1`
Role: `software`
Outcome: `findings`
Source: fresh-context subagent


## Software Review

Finding:
- [bug] `return()` no longer skips the whole installed-consumer script after the split. `tests/check_installed_package_consumer.cmake:3-4` now `include()`s `setup.cmake` and `producer-install.cmake`, but `return()` inside an included CMake file only returns control to the including file. That changes the two existing skip paths at `tests/installed-consumer-contract/setup.cmake:69-74` and `tests/installed-consumer-contract/producer-install.cmake:43-49`: missing optional required compilers and `TEST_ALLOW_CONFIGURE_SKIP` producer configure failures used to stop the script, but now the main script continues into artifact/version/source/consumer checks with a missing or failed install tree. This can turn intentional optional-path skips into hard failures, or run the contract with the wrong/default compiler after `test_fortran_compiler` was never set. The split needs a main-script-visible skip flag, a wrapper function/macro strategy, or another mechanism that preserves the original top-level `return()` behavior.

Required Questions:
1. Does this diff change public behavior? Yes. It changes CTest/package-contract behavior for optional skip paths due to CMake include/return semantics.
2. Are docs, CLI help, and tests updated together with the behavior change? Not applicable for docs/CLI if the behavior change is unintended. Tests are missing coverage for the optional skip paths above.
3. Are new defaults safe? No new defaults, but the extracted control flow is not safe for existing optional compiler/runtime checks.
4. Is any behavior silently lossy or imputed? Yes. The missing-required-compiler path can continue without setting the requested compiler, allowing later configure logic to fall back to the ambient/default compiler instead of preserving the intended skip.
5. Is there an end-to-end test path for the new feature? Partially. The normal installed-consumer paths were validated, but there is no end-to-end coverage for the skip/control-flow paths affected by the include split.

**Comment by jaharris87** [comment]
Manual fallback review record

Reviewed PR head SHA: `b0c79c6e959a9f1915900cfd7b2215e1e45d78da`
Fallback wave: `1`
Role: `test-quality`
Outcome: `findings`
Source: fresh-context subagent


## Test Quality Review

### Findings

1. **[test gap] The normal CTest pass does not prove the responsibility includes still execute.**  
   `tests/check_installed_package_consumer.cmake:13-24` now relies on side-effect includes for version probes, source rejection probes, consumer build, and consumer execution. If one of those includes were accidentally dropped, the script could still exit successfully because there is no end-of-script assertion that the expected phases ran. This matters most for `package-version-probes.cmake`, which performs all checks at include time, and `consumer-run.cmake`, which contains the report/CSV/stderr assertions. A concrete test would have each included phase append to a `ftimer_installed_consumer_contract_phases` list and have the driver assert the exact expected phase list for install-only vs full-consumer modes.

2. **[weak assertion] Source rejection probes only assert “build failed,” not “failed for the intended API contract reason.”**  
   In `tests/installed-consumer-contract/source-rejection-probes.cmake:57-67`, any nonzero build result is accepted. A syntactically broken probe source, a missing imported target, or an unexpected compiler/module issue after configure would all satisfy the rejection expectation. Add a positive-control installed-source probe that proves the harness can compile a nearby valid source, and for rejected probes assert stderr/stdout contains a stable diagnostic fragment such as the probe name plus a compiler failure around the rejected call.

3. **[test gap] MPI+OpenMP launcher and stderr diagnostics are only covered when that feature-mode test is registered and run.**  
   The stderr diagnostic preservation check lives in `tests/installed-consumer-contract/consumer-run.cmake:219-411`, but the corresponding installed-consumer mode is only registered under `FTIMER_USE_MPI AND FTIMER_USE_OPENMP` in `tests/CMakeLists.txt:299-306`. A serial smoke validation can pass while never exercising the launcher construction, timeout path, hybrid report/CSV checks, or stderr line-count assertions. The PR/follow-up should explicitly validate the split in an MPI+OpenMP smoke build, or add a script-level unit fixture that drives the launcher/stderr parsing with canned output.

### Required Questions

1. **Do the tests actually prove the behavior the PR claims to implement?**  
   Partially. The current files preserve the intended control flow, and the existing installed-consumer tests exercise the full script when the relevant feature modes are enabled. But ordinary passing CTest output alone does not prove that every side-effect include still ran.

2. **Would these tests fail for the most likely incorrect implementations? Or could a broken implementation still pass?**  
   Some broken splits would fail, especially missing setup, producer install, artifact verification, or consumer build. But dropping version probes, source rejection probes, or consumer execution could still pass unless additional phase/sentinel assertions are added.

3. **Are the important edge cases covered? What important cases are still untested?**  
   The edge cases exist in the script: install-only, absolute include dirs, DESTDIR, package versions, MPI/OpenMP modes, source rejection, launcher logic, and stderr diagnostics. The missing piece is meta-coverage that proves those checks were actually reached after the split.

4. **Are the tests deterministic and appropriately isolated?**  
   Mostly yes. The script uses per-test build/install roots and generated probe directories. The main isolation risk is environmental: MPI/OpenMP coverage depends on available compilers, runtimes, and `mpiexec`, so those branches can skip or remain unregistered outside the right build matrix.

5. **Highest-value additional tests:**  
   Add a phase-list assertion in the driver, add positive-control source probes plus intended-failure diagnostics for rejection probes, and run or gate a dedicated MPI+OpenMP installed-consumer split validation so the launcher and stderr diagnostics are not preserved only by manual inspection.

**Comment by jaharris87** [comment]
Manual fallback review record

Reviewed PR head SHA: `1c7e00ac3d1d42d449762e140a80f3846034c1a3`
Fallback wave: `2`
Role: `software`
Outcome: `no-findings`
Source: fresh-context subagent


## Software Review

No findings.

Public behavior change? No. The split keeps `tests/check_installed_package_consumer.cmake` as the ordered driver, and the repaired stop checks at `tests/check_installed_package_consumer.cmake:3-11` now preserve the prior top-level `return()` behavior for both setup/compiler-skip and producer-configure-skip paths. I also ran a missing-compiler script-mode probe; it emitted only the expected skip message and exited successfully.

Docs/tests parity? No docs changes appear needed for this helper-only extraction. Existing installed-consumer contract coverage remains in the driver flow.

Safe defaults? Yes. `ftimer_installed_consumer_contract_stop` is initialized false in `tests/installed-consumer-contract/setup.cmake:3`, set true before include-local returns in `setup.cmake:74-75` and `producer-install.cmake:48-49`, and checked immediately by the top-level driver.

Silently lossy/imputed behavior? None found. Skip behavior remains explicit via status messages; failures still fatal where they were before.

End-to-end validation? I did not rerun the full smoke/consumer suite. I reviewed the current PR head `1c7e00a`, checked the diff, ran `git diff --check main...HEAD`, and exercised the prior skip regression path with a missing required compiler. The PR body reports the relevant ctest runs passed.

**Comment by jaharris87** [comment]
Manual fallback review record

Reviewed PR head SHA: `1c7e00ac3d1d42d449762e140a80f3846034c1a3`
Fallback wave: `2`
Role: `test-quality`
Outcome: `no-findings`
Source: fresh-context subagent


## Test Quality Review

No merge-blocking test-quality findings for the current head `1c7e00a`.

The concrete regression found after the first fallback wave is addressed: the helper-level skip paths now set `ftimer_installed_consumer_contract_stop` before returning from the included file, and the driver checks that flag immediately after `setup.cmake` and `producer-install.cmake` (`tests/installed-consumer-contract/setup.cmake:67-75`, `tests/installed-consumer-contract/producer-install.cmake:43-49`, `tests/check_installed_package_consumer.cmake:3-11`). Given the coordinator’s targeted missing-required-compiler and optional-configure-skip probes, those tests now prove the most important control-flow behavior that was broken by the split.

Answers:
1. Do tests prove the claimed behavior? Yes for this organization-only split at the practical contract level. The driver still reaches install verification, package-version probes, source-rejection probes, consumer build, and consumer run in full mode (`tests/check_installed_package_consumer.cmake:13-32`), while install-only exits after artifact verification (`tests/check_installed_package_consumer.cmake:16-18`). The provided validation also covers the previously risky skip paths and MPI+OpenMP installed-consumer path.
2. Would likely broken implementations fail? Mostly yes. Missing setup, producer install, artifact verification, consumer configure/build, consumer execution, MPI launcher construction, or MPI+OpenMP report/stderr checks would fail the existing CTest paths because those helpers perform fatal assertions on missing files, bad process results, and expected report/CSV/stderr content (`tests/installed-consumer-contract/consumer-run.cmake:76-147`, `tests/installed-consumer-contract/consumer-run.cmake:169-411`). The main remaining exception is meta-coverage for accidentally dropping a side-effect-only include.
3. Edge cases? The relevant split-sensitive edge cases are covered or were explicitly revalidated: missing required compiler skip, optional configure skip, install-only absolute include/Destdir, installed consumer, build-contract checks, and MPI+OpenMP installed consumer. The MPI+OpenMP installed-consumer test is still only registered in MPI+OpenMP builds (`tests/CMakeLists.txt:299-306`), but the supplied follow-up validation covers that path for this PR.
4. Determinism/isolation? Adequate. The harness uses per-test binary/install/probe directories (`tests/installed-consumer-contract/setup.cmake:4-6`, `tests/installed-consumer-contract/source-rejection-probes.cmake:4-8`, `tests/installed-consumer-contract/package-version-probes.cmake:34-39`) and deterministic CMake/process assertions. Remaining environment dependence is expected for compiler/runtime-gated MPI/OpenMP paths, and the fixed stop flag preserves clean skips instead of false execution.
5. Highest-value additional tests? Non-blocking follow-up: add a phase-list/sentinel assertion in the driver so each helper can append an executed phase and the driver can assert the exact phase sequence for install-only vs full-consumer modes. This is a [test gap], not a blocker, because the current diff is mechanical and the actual helper includes are present at `tests/check_installed_package_consumer.cmake:13-32`.

Non-blocking follow-up observations:
- [test gap] There is still no explicit phase-list assertion proving every side-effect include ran. A future refactor could drop `package-version-probes.cmake` or `consumer-run.cmake` from `tests/check_installed_package_consumer.cmake:21-32` and potentially lose coverage without a direct “phase missing” failure.
- [weak assertion] Source rejection probes still accept any build failure after configure (`tests/installed-consumer-contract/source-rejection-probes.cmake:57-67`). A positive-control compile probe plus stable diagnostic-fragment checks would reduce false confidence, but this weakness predates the split and is not merge-blocking for this organization-only PR.
- [flaky risk/design concern] MPI/OpenMP coverage remains toolchain-gated by registration and host capabilities (`tests/CMakeLists.txt:292-330`). The supplied MPI+OpenMP validation is the right evidence for this PR; a future canned-output launcher/stderr unit fixture would make that behavior easier to prove on non-MPI hosts.

**Comment by jaharris87** [comment]
Review disposition for current head `1c7e00ac3d1d42d449762e140a80f3846034c1a3`:

- Software fallback wave 1 finding: **Agree and fixed** in commit `1c7e00ac3d1d42d449762e140a80f3846034c1a3`. The include-local `return()` regression is fixed by setting `ftimer_installed_consumer_contract_stop` before optional skip returns and checking it immediately in the driver after `setup.cmake` and `producer-install.cmake`. Validation added for both affected skip paths: missing required compiler skip and optional producer configure skip. Software fallback wave 2 found no findings on the current head.
- Test-quality wave 1 finding 1, phase-list meta-coverage: **Deferred to #331**. This is useful hardening, but non-merge-blocking for this extraction-only PR because the current helper includes are present, the split-sensitive skip regression is fixed, and the current-head follow-up review found practical contract coverage adequate.
- Test-quality wave 1 finding 2, source-rejection probes accepting any build failure: **Deferred to #331**. This weakness predates the split and tightening it would expand the tested source-rejection contract beyond issue #312 scope.
- Test-quality wave 1 finding 3, MPI+OpenMP launcher/stderr path: **Addressed with validation for this PR**. I ran `ctest --test-dir build-mpi-openmp --output-on-failure -R '^ftimer_installed_package_consumer_mpi_openmp$'`, which passed and directly exercises the hybrid launcher/report/CSV/stderr checks. #331 also tracks optional portable fixture hardening.

No merge-blocking findings remain. Manual fallback reviewer bodies for both active roles are posted for the superseded head `b0c79c6e959a9f1915900cfd7b2215e1e45d78da` and the current head `1c7e00ac3d1d42d449762e140a80f3846034c1a3`.

**Comment by jaharris87** [comment]
Codex review coverage: software covered for 1c7e00ac3d1d42d449762e140a80f3846034c1a3 via manual fallback.

**Comment by jaharris87** [comment]
Codex review coverage: test-quality covered for 1c7e00ac3d1d42d449762e140a80f3846034c1a3 via manual fallback.

File: tests/installed-consumer-contract/setup.cmake [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+endif()
+
+if(DEFINED TEST_REQUIRED_COMPILER_NAMES AND NOT TEST_REQUIRED_COMPILER_NAMES STREQUAL "")
+  string(REPLACE "," ";" required_compiler_names "${TEST_REQUIRED_COMPILER_NAMES}")
+  find_program(ftimer_required_compiler NAMES ${required_compiler_names})
+  if(NOT ftimer_required_compiler)
+    message(STATUS
+      "Skipping ${test_name}: none of the required compilers are available on PATH (${TEST_REQUIRED_COMPILER_NAMES})."
+    )
+    return()
```
[the remark this answers is not in the document]
Comment (jaharris87) [reply tests/installed-consumer-contract/setup.cmake:75]: Fixed in `1c7e00ac3d1d42d449762e140a80f3846034c1a3` by setting `ftimer_installed_consumer_contract_stop` before this include-local `return()` and checking that flag in the top-level driver immediately after `setup.cmake` is included. I also validated the missing-required-compiler skip path with a script-mode probe.

File: tests/installed-consumer-contract/producer-install.cmake [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+execute_process(
+  COMMAND "${CMAKE_COMMAND}" ${configure_args}
+  RESULT_VARIABLE producer_configure_result
+)
+if(NOT producer_configure_result EQUAL 0)
+  if(TEST_ALLOW_CONFIGURE_SKIP)
+    message(STATUS
+      "Skipping ${test_name}: producer configure failed for the optional compiler/runtime path."
+    )
+    return()
```
[the remark this answers is not in the document]
Comment (jaharris87) [reply tests/installed-consumer-contract/producer-install.cmake:49]: Fixed in `1c7e00ac3d1d42d449762e140a80f3846034c1a3` by setting `ftimer_installed_consumer_contract_stop` before the allowed producer-configure skip return and checking that flag in the top-level driver immediately after `producer-install.cmake` is included. I also validated this optional-configure-skip path with a script-mode probe.
